htlcswitch/mailbox.go

```diff
@@ -365,206 +365,190 @@ func (p *pktWithExpiry) deadline(clock clock.Clock) <-chan time.Time {
-	switch cType {
-	case wireCourier:
-		defer close(m.wireShutdown)
-	case pktCourier:
-		defer close(m.pktShutdown)
-	}
-
-	// TODO(roasbeef): refactor...
+// wireMailCourier is a dedicated goroutine whose job is to reliably deliver
+// wire messages.
+func (m *memoryMailBox) wireMailCourier() {
+	defer close(m.wireShutdown)
 
 	for {
-		// First, we'll check our condition. If our target mailbox is
-		// empty, then we'll wait until a new item is added.
-		switch cType {
-		case wireCourier:
-			m.wireCond.L.Lock()
-			for m.wireMessages.Front() == nil {
-				m.wireCond.Wait()
-
-				select {
-				case msgDone := <-m.msgReset:
-					m.wireMessages.Init()
-
-					close(msgDone)
-				case <-m.quit:
-					m.wireCond.L.Unlock()
-					return
-				default:
-				}
-			}
+		// First, we'll check our condition. If our mailbox is empty,
+		// then we'll wait until a new item is added.
+		m.wireCond.L.Lock()
+		for m.wireMessages.Front() == nil {
+			m.wireCond.Wait()
 
-		case pktCourier:
-			m.pktCond.L.Lock()
-			for m.repHead == nil && m.addHead == nil {
-				m.pktCond.Wait()
-
-				select {
-				// Resetting the packet queue means just moving
-				// our pointer to the front. This ensures that
-				// any un-ACK'd messages are re-delivered upon
-				// reconnect.
-				case pktDone := <-m.pktReset:
-					m.repHead = m.repPkts.Front()
-					m.addHead = m.addPkts.Front()
-
-					close(pktDone)
-
-				case <-m.quit:
-					m.pktCond.L.Unlock()
-					return
-				default:
-				}
+			select {
+			case msgDone := <-m.msgReset:
+				m.wireMessages.Init()
+				close(msgDone)
+			case <-m.quit:
+				m.wireCond.L.Unlock()
+				return
+			default:
 			}
 		}
 
-		var (
-			nextRep   *htlcPacket
-			nextRepEl *list.Element
-			nextAdd   *pktWithExpiry
-			nextAddEl *list.Element
-			nextMsg   lnwire.Message
-		)
-		switch cType {
 		// Grab the datum off the front of the queue, shifting the
 		// slice's reference down one in order to remove the datum from
 		// the queue.
-		case wireCourier:
-			entry := m.wireMessages.Front()
-			nextMsg = m.wireMessages.Remove(entry).(lnwire.Message)
+		entry := m.wireMessages.Front()
 
-		// For packets, we actually never remove an item until it has
-		// been ACK'd by the link. This ensures that if a read packet
-		// doesn't make it into a commitment, then it'll be
-		// re-delivered once the link comes back online.
-		case pktCourier:
-			// Peek at the head of the Settle/Fails and Add queues.
-			// We peak both even if there is a Settle/Fail present
-			// because we need to set a deadline for the next
-			// pending Add if it's present. Due to clock
-			// monotonicity, we know that the head of the Adds is
-			// the next to expire.
-			if m.repHead != nil {
-				nextRep = m.repHead.Value.(*htlcPacket)
-				nextRepEl = m.repHead
-			}
-			if m.addHead != nil {
-				nextAdd = m.addHead.Value.(*pktWithExpiry)
-				nextAddEl = m.addHead
-			}
-		}
+		//nolint:forcetypeassert
```

Comment:
My only reservation with this PR is the addition of these directives instead of the proper handling code. Maybe it's the case that this PR shouldn't be responsible for making semantics changes in addition to the refactor, but I wonder when it will be done and how brittle other parts of this code are because these are here. I do like that you include these directives in this commit but perhaps we should add a follow-on commit to this PR that "fixes this properly".

Reply:
Yeah I think we can remove the `//nolint:forcetypeassert` and fix it in a following commit since it's fairly small.

Reply:
What do you propose we do instead of panicking on incorrect type?

IMO the best solution is to use a generic typed list instead of `container/list`.  Then the type casting could be eliminated completely and incorrect usage would give a compile-time error instead of runtime panic.  But I'm not aware of a generic list implementation in the standard library.  Panicking on incorrect usage seems like the next best thing to me, as it would quickly notify us if we introduce a bug into `AddMessage`, `AddPacket`, `wireMailCourier`, or `pktMailCourier`.

Reply:
> panicking on incorrect usage seems like the next best thing to me, as it would quickly notify us if we introduce a bug into AddMessage, AddPacket, wireMailCourier, or pktMailCourier.

Yeah that sounds good. Was thinking whether it would have more heap escapes but I guess it's fine since we are already using interface values here anyway. I think we can keep the PR as is and add a todo note here and should be good to go. Then we can refactor it in a new PR.

Reply:
Added a TODO in https://github.com/lightningnetwork/lnd/pull/7846/commits/b1a3c46759be0c5b1492c15a46bf81905e475a3e.